Integration/azd + pre-commit

JinLee794 opened the pull request:
# Project Enhancement Request - #3 

## Description
This pull request contains several updates across different files, focusing on workflow automation, environment configuration, and code cleanup. The most important changes include the addition of a new GitHub Actions workflow for deploying to Azure, updates to existing workflows and configurations, and the removal of pre-commit hooks for code linting.

### Workflow and Deployment Updates:
* [`.github/workflows/azd_deploy.yml`](diffhunk://#diff-dc2a4bfb4733c1efa6a5941e194daa6a2952f0334817d89f62c5c17653675751R1-R72): Added a new GitHub Actions workflow to deploy to Azure using AZD, including steps for provisioning infrastructure and deploying the application.
* [`.github/workflows/main_staging_ci.yml`](diffhunk://#diff-ae5ff18a16f52ba98bef360d59f78abf1da112e336db8a32b2a46bf1cda11c47L20-R21): Updated the Conda setup step to use `environment.yaml` and removed the `activate-environment` value.

### Configuration Changes:
* [`.pre-commit-config.yaml`](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9L4-R30): Commented out the pre-commit hooks for `flake8`, `mypy`, and `ruff` to temporarily disable code linting.
* [`.streamlit/config.toml`](diffhunk://#diff-a7f7962e3e1209cfc31a2d3faee8dd8fdd5341f7249033e948c6539eae0a47c3L1-L9): Removed the Streamlit configuration settings for theme and server.
* [`.vscode/launch.json`](diffhunk://#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945R4-R13): Added a new launch configuration for debugging Streamlit applications in VS Code.

### Directory Structure Changes:
- Removed the `devops` directory
	- moved the infra directory to root (`devops/infra` -> `infra`)
	- moved the Dockerfile from the `devops/container` directory to the root directory
	- removed the docker-compose, as it was only being leveraged to set the directory `context` (now handled by [azure.yaml](azure.yaml)

### Infra Code Changes:

#### AI Search
- Now leverages system managed identity to authenticate to blob storage for data source (fixes storage key-based auth issue)
	-  see [infra/modules/data/search.bicep]("infra/modules/data/search.bicep")
- Restructured the bicep deployment to fit into the `azd` mold
	- created a `subscription` scoped `main.bicep`, which invokes the resource-group scoped deployment for priorauth.
	- added additional output variables on `main.bicep` and the `priorAuthInfra.bicep` to allow for azd to capture the output values to then translate to a .env file for local development purposes (orchestrated via [azure.yaml](azure.yaml))

#### Container App
- Now supports optional user-assigned-identity to be provided
- Currently updated the priorAuthInfra to create a user assigned identity, grant it storage blob reader + data contrib, and pass it into the container app module to further associate with the container app job and the container app, while granting it additional permissions to manage container app jobs.
- Container App Job only gets initialized through the bicep with a quickstart image. The postdeploy step in the azure.yaml handles the triggering of the job.

### Code Refactoring for RBAC:
- Storage client ([blob_helper.py](src/storage/blob_helper.py)) now checks if the connection string contains a `ResourceId=` string as that implies to default to using a managed identity/user assigned identity to communicate
- Similar changes on src/pipeline/paprocessing/run.py

### Code Cleanup:
* [`app/components/managers.py`](diffhunk://#diff-cb3ac37c881dc8c5fe2affe10cfe545a7fd97114e796650cffdc7c70cf53dae0L1-L4): Removed unused imports to clean up the code.
* [`app/streamlit/Home.py`](diffhunk://#diff-6eb2a3202757db68a6b86cac933ae1e13e699c4b9b87ef419e0d308239af3085L6-R6): Removed unused imports and simplified the import statements.

### Notebook Updates:
* [`01-indexing-policies.ipynb`](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L5-R12): Cleared the execution counts and outputs for all code cells to reset the notebook state. [[1]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L5-R12) [[2]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L78-R72) [[3]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L106-R88) [[4]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L130-R104) [[5]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L154-R120) [[6]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L178-R136) [[7]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L220-R170) [[8]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L294-R232) [[9]](diffhunk://#diff-850469b58b43e784551eb23c167fa2db7b2aad52411fb191a486ae22267475a7L389-R283)

### Dockerfile Updates:
* `Dockerfile` (renamed from `devops/container/Dockerfile`): Updated the base image to `python:3.11-slim-bullseye`, added `python3-dev` to the list of installed packages, and cleaned up the file. [[1]](diffhunk://#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L1-R1) [[2]](diffhunk://#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L24-R24) [[3]](diffhunk://#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L35)

### Documentation Updates:
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L12-R12): Fixed the link for the "Deploy To Azure" button to point to the correct JSON file for deployment.
Fixes #(issue) <!-- Mention the related issue if applicable -->

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have successfully executed and passed all new and existing tests and checks relevant to the project.
- [x] I have updated the documentation if necessary.
- [x] I have tested the changes in an environment representative of the deployment environment.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have opened an Issue reporting a Bug, New Feature, Breaking Change, and/or Documentation Update (https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue).

## Azure AI Service Integration Details (if applicable):

- [ ] I have integrated a new Azure AI service.
- [x] I have updated the configuration or usage of an existing Azure AI service.
- [ ] The integration has been tested and documented.
- [ ] Changes to the integration are backwards-compatible with previous versions.

## Context

Provide additional context on the changes you are proposing, the problem they address, and their significance for the project.

marcjimz commented:
@JinLee794 huge PR! I am around to give it a whirl + deploy, can tag me later this week!

pablosalvador10 commented:
@JinLee794 awesome Job !!

JinLee794 commented:
@pablosalvador10 @marcjimz - I pulled in the changes from #24 and adjusted the deployment accordingly. I also replaced the containerapp.bicep module with direct references to the avm modules inside the priorAuth.bicep.
